tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs

```diff
@@ -102,253 +143,435 @@ private void SetGeneralConfig(TracerSettings tracerSettings, DebuggerSettings se
             ServiceName = GetServiceName(tracerSettings);
         }
 
-        private string GetServiceName(TracerSettings tracerSettings)
+        internal Task UpdateConfiguration(TracerSettings tracerSettings, DebuggerSettings? newDebuggerSettings = null)
         {
-            try
+            return UpdateProductsState(tracerSettings, newDebuggerSettings ?? DebuggerSettings);
+        }
+
+        private Task UpdateProductsState(TracerSettings tracerSettings, DebuggerSettings newDebuggerSettings)
+        {
+            if (_processExit.Task.IsCompleted)
             {
-                return TraceUtil.NormalizeTag(tracerSettings.ServiceName ?? TracerManager.Instance.DefaultServiceName);
+                return _processExit.Task;
             }
-            catch (Exception e)
+
+            OneTimeSetup(tracerSettings);
+
+            lock (_syncLock)
             {
-                Log.Error(e, "Could not set `DynamicInstrumentationHelper.ServiceName`.");
-                return TracerManager.Instance.DefaultServiceName;
+                if (_processExit.Task.IsCompleted)
+                {
+                    return _processExit.Task;
+                }
+
+                DebuggerSettings = newDebuggerSettings;
+                SetCodeOriginState(newDebuggerSettings);
+                SetExceptionReplayState(newDebuggerSettings);
+            }
+
+            return DebouncedUpdateDynamicInstrumentationAsync(tracerSettings, newDebuggerSettings);
+        }
+
+        private void OneTimeSetup(TracerSettings tracerSettings)
+        {
+            if (Interlocked.CompareExchange(ref _initialized, 1, 0) != 0)
+            {
+                return;
+            }
+
+            LifetimeManager.Instance.AddShutdownTask(ShutdownTasks);
+            SetGeneralConfig(tracerSettings, DebuggerSettings);
+            InitializeSymbolUploader(tracerSettings);
```

Comment:
I forgot to mention, discussing with @tylfin he pointed out that the symbol uploading should _not_ be enabled by default, so I think we need to rework this to make sure symbol uploading is lazy, and doesn't start until DI is enabled (i.e. immediately if DI is explicitly enabled, or not until it's enabled remotely)

The concern is that this is a lot of additional network traffic on both sides (client + backend) if it's not going to be used